### PR TITLE
patch: bump typescript and fix TS errors

### LINKED ISSUE #5176: bug: TRPCClientError.ts compilation error due to typescript 5.1.3 type narrowing regression
### Provide environment information

```
  System:
    OS: Linux 6.6 Arch Linux
    CPU: (16) x64 AMD Ryzen 7 7840U w/ Radeon  780M Graphics
    Memory: 19.53 GB / 30.55 GB
    Container: Yes
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.10.0 - ~/.nvm/versions/node/v20.10.0/bin/node
    Yarn: 1.22.21 - ~/.nvm/versions/node/v20.10.0/bin/yarn
    npm: 10.2.3 - ~/.nvm/versions/node/v20.10.0/bin/npm
    pnpm: 8.12.0 - ~/.nvm/versions/node/v20.10.0/bin/pnpm
  Browsers:
    Brave Browser: 120.1.61.101
  npmPackages:
    @trpc/client: ^10.25.0 => 10.44.1
    @trpc/server: ^10.25.0 => 10.44.1
    typescript: 5.2.2 => 5.2.2
```

### Describe the bug

This is a bug related to this PR: https://github.com/trpc/trpc/pull/4796

There is a type narrowing regression present in `5.1.3` , the typescript version currently used in the trpc project, that is fixed in `5.1.5` and newer versions. When using trpc in a project with a newer typescript version where the regression is fixed a type issue is present. The regression is causing the following error in `TRPCClientError.ts`:

```
node_modules/@trpc/client/src/TRPCClientError.ts:103:7 - error TS2322: Type 'TRPCClientError<ErrorInferrable>' is not assignable to type 'TRPCClientError<TRouterOrProcedure>'.
  Type 'ErrorInferrable' is not assignable to type 'TRouterOrProcedure'.
    'ErrorInferrable' is assignable to the constraint of type 'TRouterOrProcedure', but 'TRouterOrProcedure' could be instantiated with a different subtype of constraint 'ErrorInferrable'.
      Type 'AnyProcedure' is not assignable to type 'TRouterOrProcedure'.
        'AnyProcedure' is assignable to the constraint of type 'TRouterOrProcedure', but 'TRouterOrProcedure' could be instantiated with a different subtype of constraint 'ErrorInferrable'.

  101 |     }
  102 |     if (isTRPCErrorResponse(cause)) {
> 103 |       return new TRPCClientError(cause.error.message, {
      |       ^
  104 |         ...opts,
  105 |         result: cause,
  106 |       });
```

Installing a newer version of typescript than `5.1.3`, or downgrading to an earlier version (`5.0.4` is the latest stable version before `5.1.3`) reproduces this error.

Issues fixed in the `5.1.5` typescript version: https://github.com/Microsoft/TypeScript/issues?utf8=%E2%9C%93&q=milestone%3A%22TypeScript+5.1.5%22+is%3Aclosed+

I believe this is the specific PR to fix the regression (linked issue with details in the PR description): https://github.com/microsoft/TypeScript/pull/54052

### Link to reproduction

https://github.com/jeremyhofer/trpc-typescript-version-compiler-issues

### To reproduce

Simply changing the typescript version from `5.1.3` to a different version, such as a newer version of `5.1.x` or downgrading to a `5.0.x` version, reproduces this issue.

I was having trouble reproducing in the example stackblitz environment provided, probably due to configuration and build differences between that example and how my and other projects are configured. However, the errors are reproducible in the trpc repository itself. I have linked a fork with a commit changing the typescript version to `5.1.6` and setting `noPropertyAccessFromIndexSignature` to `true` in `tsconfig.build.json` (this 2nd change done for the related issue in #5175). Changing the typescript version is sufficient enough to reproduce the issue in the trpc build when attempting to `pnpm build` after a fresh `pnpm i`.

### Additional information

Related to #5175 .

Related issue from AnalogJS where myself and others have been experiencing this issue: https://github.com/analogjs/analog/issues/722

Also confirmed that downgrading `@trpc/client` to `10.38.2`, the version before https://github.com/trpc/trpc/pull/4796 was released, resolves both problems as the code in question is not present in that version.

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/5176">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/5176/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by KATT**
Closes #5176

## 🎯 Changes

Upgrades TS and fixes errors.

### CODE REVIEW

**Comment by jeremyhofer** [comment]
@KATT managed to address all the compilation errors after the version bump. I fixed the type issues as best I could given my being new to the internals of trpc, as well as my experience with (extremely!) advanced type definitions/usage being limited. I imagine there are alternatives possible. Let me know if you want me to try and tackle things in a different way at all.

File: packages/server/tsconfig.json [hunk 1]
```diff
@@ -5,6 +5,7 @@
     // https://github.com/tsconfig/bases/blob/main/bases/node18.json
     "lib": ["es2023", "DOM", "DOM.Iterable"],
     "target": "es2022",
-    "module": "node16"
+    "module": "Node16",
+    "moduleResolution": "node16"
```
Comment (juliusmarminge) [inline packages/server/tsconfig.json:9]: we want Bundler i think? given that we transpile with rollup

File: packages/tests/server/react/polymorphism.test.tsx [hunk 2]
```diff
@@ -458,12 +458,12 @@ function RefreshExportsListButton({
...
 }
 
-type ExportStatusProps<TStatus extends Factory.ExportRouteLike['status']> = {
+type ExportStatusProps<TStatus extends QueryLike<AnyProcedure>> = {
   status: TStatus;
   renderAdditionalFields?: (data: InferQueryLikeData<TStatus>) => ReactNode;
   currentExport: number | null;
 };
-function ExportStatus<TStatus extends Factory.ExportRouteLike['status']>({
+function ExportStatus<TStatus extends QueryLike<AnyProcedure>>({
```
Comment (KATT) [inline packages/tests/server/react/polymorphism.test.tsx:466]: cc @Nick-Lucas, could you help here? 👀

Comment (Nick-Lucas) [reply packages/tests/server/react/polymorphism.test.tsx]: Pushed, actually a simple change

File: packages/server/tsconfig.json (hunk 1 above)
Comment (KATT) [reply packages/server/tsconfig.json:9]: Not sure tbh - it just started complaining and I made it stop with this change

File: www/package.json [hunk 3]
```diff
@@ -89,7 +89,7 @@
     "tsx": "^4.0.0",
     "typedoc": "^0.24.7",
     "typedoc-plugin-markdown": "^4.0.0-next.13",
-    "typescript": "^5.1.3",
+    "typescript": "npm:typescript@~5.1.3",
```
Comment (KATT) [inline www/package.json:92]: couldn't get docusaurus to compile with 5.3.3 🤷

**Pull request merged**